File: FrontEnd/lib/domain/entities/reminder.dart
/// @file reminder.dart
/// @description Entidad de dominio Reminder. Representa un recordatorio de cuidado de planta.
/// Objeto puro de Dart, sin dependencias de Flutter ni paquetes externos.
/// Los mappers convierten ReminderModel ↔ Reminder.
/// @module Reminders
/// @layer Domain
library;

// ═══════════════════════════════════════════════════════════════════════════════
// REMINDER ENTITY
// ═══════════════════════════════════════════════════════════════════════════════

/// Entidad de dominio que representa un recordatorio de cuidado de planta.
///
/// Todos los campos son inmutables (final). Usar [copyWith] para actualizaciones.
class Reminder {
  const Reminder({
    required this.id,
    required this.plantId,
    required this.userId,
    required this.type,
    required this.scheduledDate,
    required this.message,
    required this.isCompleted,
    required this.suspended,
    required this.attempts,
    required this.createdAt,
  });

  /// Identificador único del recordatorio (MongoDB ObjectId como String).
  final String id;

  /// ID de la planta asociada al recordatorio.
  final String plantId;

  /// ID del usuario propietario.
  final String userId;

  /// Tipo de recordatorio: 'watering', 'fertilizing', 'pruning', 'repotting', 'other'.
  final String type;

  /// Fecha y hora programada para el recordatorio (UTC).
  final DateTime scheduledDate;

  /// Mensaje descriptivo del recordatorio (p.ej. "Regar la planta").
  final String message;

  /// true si el recordatorio ha sido completado por el usuario.
  final bool isCompleted;

  /// true si el recordatorio está suspendido (p.ej. por el ajuste climático).
  final bool suspended;

  /// Número de intentos de procesamiento realizados por el cron job.
  final int attempts;

  /// Fecha de creación del recordatorio.
  final DateTime createdAt;

  // ─── Helpers ─────────────────────────────────────────────────────────────────

  /// true si el recordatorio está pendiente de procesar (no completado ni suspendido).
  bool get isPending => !isCompleted && !suspended;

  /// true si el recordatorio es de tipo riego.
  bool get isWatering => type == 'watering';

  /// true si el recordatorio está vencido (scheduledDate en el pasado y aún pendiente).
  bool get isOverdue =>
      isPending && scheduledDate.isBefore(DateTime.now().toUtc());

  // ─── copyWith ────────────────────────────────────────────────────────────────

  Reminder copyWith({
    String?   id,
    String?   plantId,
    String?   userId,
    String?   type,
    DateTime? scheduledDate,
    String?   message,
    bool?     isCompleted,
    bool?     suspended,
    int?      attempts,
    DateTime? createdAt,
  }) {
    return Reminder(
      id:            id            ?? this.id,
      plantId:       plantId       ?? this.plantId,
      userId:        userId        ?? this.userId,
      type:          type          ?? this.type,
      scheduledDate: scheduledDate ?? this.scheduledDate,
      message:       message       ?? this.message,
      isCompleted:   isCompleted   ?? this.isCompleted,
      suspended:     suspended     ?? this.suspended,
      attempts:      attempts      ?? this.attempts,
      createdAt:     createdAt     ?? this.createdAt,
    );
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) || (other is Reminder && other.id == id);

  @override
  int get hashCode => id.hashCode;

  @override
  String toString() =>
      'Reminder(id: $id, type: $type, plantId: $plantId, isPending: $isPending)';
}

File: BackEnd/src/domain/entities/Reminder.ts
/**
 * @file Reminder.ts
 * @description Entidad de dominio que representa un recordatorio de cuidado de planta.
 * Los recordatorios de tipo watering/pruning se reprograman automáticamente tras procesarse.
 * Los de tipo custom son de uso único y se marcan como completados tras el primer envío.
 * @module Reminders
 * @layer Domain
 */

/** Tipo de recordatorio */
export type ReminderType = 'watering' | 'pruning' | 'custom';

/**
 * Entidad Reminder.
 * Representa una tarea programada de cuidado asociada a una planta y usuario.
 */
export class Reminder {
  readonly id:            string;
  /** Id de la planta a la que pertenece este recordatorio */
  readonly plantId:       string;
  /** Id del usuario propietario */
  readonly userId:        string;
  /** Tipo de cuidado: riego, poda o personalizado */
  readonly type:          ReminderType;
  /** Fecha/hora en que el recordatorio debe procesarse */
  readonly scheduledDate: Date;
  /** Mensaje de la notificación push */
  readonly message:       string;
  /** true si se ha completado definitivamente (reminders de tipo custom) */
  readonly isCompleted:   boolean;
  /** true si se ha suspendido temporalmente (ej.: por lluvia prevista) */
  readonly suspended:     boolean;
  /** Número de veces que se ha intentado enviar la notificación */
  readonly attempts:      number;
  readonly createdAt:     Date;

  constructor(params: {
    id:            string;
    plantId:       string;
    userId:        string;
    type:          ReminderType;
    scheduledDate: Date;
    message:       string;
    isCompleted:   boolean;
    suspended:     boolean;
    attempts:      number;
    createdAt:     Date;
  }) {
    this.id            = params.id;
    this.plantId       = params.plantId;
    this.userId        = params.userId;
    this.type          = params.type;
    this.scheduledDate = params.scheduledDate;
    this.message       = params.message;
    this.isCompleted   = params.isCompleted;
    this.suspended     = params.suspended;
    this.attempts      = params.attempts;
    this.createdAt     = params.createdAt;
  }

  /**
   * true si el recordatorio está pendiente de procesar (vencido, no completado, no suspendido).
   */
  get isPending(): boolean {
    return !this.isCompleted && !this.suspended && this.scheduledDate <= new Date();
  }
}
